Libraries/LibWeb/CSS/SelectorEngine.cpp

```diff
@@ -759,6 +760,140 @@ static inline bool matches_pseudo_class(CSS::Selector::SimpleSelector::PseudoCla
+            if (form_associated_element.is_candidate_for_constraint_validation() && !form_associated_element.satisfies_its_constraints())
+                return true;
+        }
+
+        // - form elements that are the form owner of one or more elements that themselves are candidates for constraint validation but do not satisfy their constraints
+        if (is<Web::HTML::HTMLFormElement>(element)) {
+            auto& form_element = as<Web::HTML::HTMLFormElement>(element);
+            bool has_invalid_elements = false;
+            element.for_each_in_subtree([&](auto& node) {
+                if (auto const* form_associated_element = dynamic_cast<Web::HTML::FormAssociatedElement const*>(&node)) {
+                    if (form_associated_element->form() == &form_element && form_associated_element->is_candidate_for_constraint_validation() && !form_associated_element->satisfies_its_constraints()) {
+                        has_invalid_elements = true;
+                        return TraversalDecision::Break;
+                    }
+                }
+                return TraversalDecision::Continue;
+            });
+            if (has_invalid_elements)
+                return true;
+        }
+
+        // - fieldset elements that have of one or more descendant elements that themselves are candidates for constraint validation but do not satisfy their constraints
+        if (is<Web::HTML::HTMLFieldSetElement>(element)) {
+            bool has_invalid_children = false;
+            element.for_each_in_subtree([&](auto& node) {
+                if (auto const* form_associated_element = dynamic_cast<Web::HTML::FormAssociatedElement const*>(&node)) {
+                    if (form_associated_element->is_candidate_for_constraint_validation() && !form_associated_element->satisfies_its_constraints()) {
+                        has_invalid_children = true;
+                        return TraversalDecision::Break;
+                    }
+                }
+                return TraversalDecision::Continue;
+            });
+            if (has_invalid_children)
+                return true;
+        }
+
+        return false;
+    }
+    case CSS::PseudoClass::UserValid: {
+        // https://html.spec.whatwg.org/#selector-user-valid
+        // The :user-valid pseudo-class must match input, textarea, and select elements whose user validity is true,
+        bool user_validity = false;
+        if (is<Web::HTML::HTMLInputElement>(element)) {
+            user_validity = static_cast<Web::HTML::HTMLInputElement const&>(element).user_validity();
+        } else if (is<Web::HTML::HTMLSelectElement>(element)) {
+            user_validity = static_cast<Web::HTML::HTMLSelectElement const&>(element).user_validity();
+        } else if (is<Web::HTML::HTMLTextAreaElement>(element)) {
+            user_validity = static_cast<Web::HTML::HTMLTextAreaElement const&>(element).user_validity();
```

Comment:
Nit: These could use `as_if` instead of `is` and a separate cast.

Reply:
Done.